contrib/babelfishpg_common/src/datetime.c

```diff
@@ -679,6 +682,596 @@ datetime_pl_datetime(PG_FUNCTION_ARGS)
+						if (regcomp(&time_regex, pattern, REG_EXTENDED) != 0)
+							ereport(ERROR,
+										(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("time format internal error")));
+						if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+							TIME_IN_ERROR();
+						regfree(&time_regex);
+						res = makeStringInfo();
+						/*
+						 * For inputs of format "HH:MM:SS:nnn" will be converted to "HH:MM:SS.nnn".
+						 * where nnn ranges from 0 to 999.
+						 */
+						for(j = 0; j <len; j++)
+						{
+							if(field[i][j]== ':')
+							{
+								time_count++;
+								if(time_count == 3)
+								{
+									field[i][j]='.';
+								}
+							}
+							if(time_count == 3 && field[i][j]!= '.')
+							{
+								appendStringInfo(res, "%c", field[i][j]);
+							}
+						}
+						temp_time = res->data;
+						pfree(res);
+						if(time_count >2)
+						{
+							temp_field = pstrdup(field[i]);
+							res = makeStringInfo();
+							appendStringInfo(res, "%s", strtok(temp_field, "."));
+							len = strlen(temp_time);
+							/*
+							 * If nnn if of length 1 then convert to '.00n'
+							 * else if lenght is 2 then convert to '.0nn'
+							 * else if lenght is >3 then an error should be thrown
+							 */
+							if(len == 1)
+							{
+								appendStringInfo(res, "%s%s%s", ".", "00", temp_time);
+								field[i]=res->data;
+							}
+							else if(len == 2)
+							{
+								appendStringInfo(res, "%s%s%s", ".", "0", temp_time);
+								field[i]=res->data;
+							}
+							else if(len != 3)
+								TIME_IN_ERROR();
+							pfree(res);
+						}
+					}
+					break;
+
+				case DTK_STRING:
+					if (pg_strcasecmp(field[i], "z") == 0)
+					{
+						/*
+						 * If the modified_str contains the 'z' in it for example '<TIME> Z'
+						 * then the 'z' present in it should be 'Z'.
+						 */
+						pattern = "Z";
+						if (regcomp(&time_regex, pattern, 0) != 0)
+							ereport(ERROR,
+										(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("time format internal error")));
+						if (regexec(&time_regex, modified_str, 0, NULL, 0))
+							TIME_IN_ERROR();
+						regfree(&time_regex);
+					}
+					if (pg_strcasecmp(field[i], "am") == 0 || pg_strcasecmp(field[i], "pm") == 0)
+					{
+						/*
+						 * If "am/pm" is present in the input then the previous field cannot be 'z'.
+						 * For example "1 pm Z" is valid but "1 Z pm" is invalid.
+						 */
+						if(i-1 >= 0 && ftype[i-1] == DTK_STRING &&  pg_strcasecmp(field[i-1], "z") == 0)
+							TIME_IN_ERROR();
+					}
+					break;
+
+				case DTK_TZ:
+					/*
+					 * If the input contains timezone then the format should be "[+-]ZZ:ZZ" and the previous field should be time.
+					 */
+					if(i-1 < 0 || (i-1 >= 0 && ftype[i-1] != DTK_TIME))
+						TIME_IN_ERROR();
+					/*
+					 * For example error will be thrown if the field is "[+-]02",
+					 * and no error will be thrown if field is "[+-]02:00"
+					 */
+					pattern = "([0-9]?[0-9]:[0-9]?[0-9])$";
```

Comment:
where are we validating that only +/- are before offset? on PG side?

Reply:
Yes, in the function `ParseDateTime` the field is determined as `DTK_TZ` only if the input contains {+-} in-front.